### DISCUSSION TITLE
XAML `CommandParameter` and `enum` values

### DISCUSSION THREAD

**Discussion by hawkerm** [CommunityToolkit/dotnet#407] [Q&A] [answered]
Feel like this may be a bug (or oversight), but I'm also not sure how to work around it...

**Background:** Came about as I was taking the [WTS](https://aka.ms/wts) template for Settings Page which uses the `ElementTheme` XAML type in **UWP** and saves it as a property in the Settings ViewModel. I wanted to de-couple the UI from the ViewModel, as I have my MVVM bits in a .NET Standard library 😇; so I made my own `ElementTheme` enum to represent the same values:

```cs
public enum ElementTheme
{
    Default = 0,
    Light = 1,
    Dark = 2
}
```

This is exposed in a command in the view model like so:

```cs
    [RelayCommand]
    public async Task SwitchTheme(MyLib.ElementTheme newTheme)
    {
        ElementTheme = newTheme;
        await ThemeSelectorService.SetThemeAsync(newTheme);
    }
```

And then used directly in XAML here in the application layer:

```xml
                    <RadioButton
                        x:Uid="Settings_Theme_Light"
                        Command="{x:Bind ViewModel.SwitchThemeCommand}"
                        IsChecked="{x:Bind ViewModel.ElementTheme, Converter={StaticResource EnumToBooleanConverter}, ConverterParameter=Light, Mode=OneWay}"
                        FontSize="15"
                        GroupName="AppTheme">
                        <RadioButton.CommandParameter>
                            <models:ElementTheme>Light</models:ElementTheme> <!-- This used to be xaml: in the original WTS version as it directly used the XAML based enum version -->
                        </RadioButton.CommandParameter>
                    </RadioButton>
```

However, it seems like the enum is being converted by XAML or WinRT to an Int32 between the layers, and now I'm getting an exception from the `AsyncRelayCommand` that it is not a valid parameter type (even though it should be able to convert between the two):

> System.ArgumentException
  HResult=0x80070057
  Message=Parameter "parameter" (object) cannot be of type System.Int32, as the command type requires an argument of type MyLib.ElementTheme.
Parameter name: parameter
  Source=CommunityToolkit.Mvvm
  StackTrace:
   at CommunityToolkit.Mvvm.Input.RelayCommand`1.ThrowArgumentExceptionForInvalidCommandArgument(Object parameter) in /_/CommunityToolkit.Mvvm/Input/RelayCommand{T}.cs:line 170

https://github.com/CommunityToolkit/dotnet/blob/9d1a525d04299e5421cb3e5917825fee2bc11a18/CommunityToolkit.Mvvm/Input/RelayCommand%7BT%7D.cs#L167-L170

Being called from here L271:

https://github.com/CommunityToolkit/dotnet/blob/9d1a525d04299e5421cb3e5917825fee2bc11a18/CommunityToolkit.Mvvm/Input/AsyncRelayCommand%7BT%7D.cs#L263-L271

It's happening when the command is getting set by the generated xaml from the binding `obj.Command = `:

```cs
            public static void Set_Windows_UI_Xaml_Controls_Primitives_ButtonBase_Command(global::Windows.UI.Xaml.Controls.Primitives.ButtonBase obj, global::System.Windows.Input.ICommand value, string targetNullValue)
            {
                if (value == null && targetNullValue != null)
                {
                    value = (global::System.Windows.Input.ICommand) global::Windows.UI.Xaml.Markup.XamlBindingHelper.ConvertValue(typeof(global::System.Windows.Input.ICommand), targetNullValue);
                }
                obj.Command = value;
            }
```

Which somehow triggers a `CanExecute` call in the transition somewhere, I couldn't dig into that, but here's the `CanExecute` method in the stack from the `ICommandToWinRTAdapter` class:

```cs
		private bool CanExecute(object parameter)
		{
			ICommand _this = Unsafe.As<ICommand>(this);
			return _this.CanExecute(parameter);
		}
```

Call stack from there:

``` 	
CommunityToolkit.Mvvm.dll!CommunityToolkit.Mvvm.Input.RelayCommand<MyLib.ElementTheme>.ThrowArgumentExceptionForInvalidCommandArgument(object parameter) Line 170	C#
CommunityToolkit.Mvvm.dll!CommunityToolkit.Mvvm.Input.AsyncRelayCommand<MyLib.ElementTheme>.CanExecute(object parameter) Line 271	C#
System.Runtime.WindowsRuntime.dll!System.Runtime.InteropServices.WindowsRuntime.ICommandToWinRTAdapter.CanExecute(object parameter) Line 39	C#
[Native to Managed Transition]	
[Managed to Native Transition]	
MyApp.exe!MyApp.Views.SettingsPage.XamlBindingSetters.Set_Windows_UI_Xaml_Controls_Primitives_ButtonBase_Command(Windows.UI.Xaml.Controls.Primitives.ButtonBase obj, System.Windows.Input.ICommand value, string targetNullValue) Line 36	C#
```

🤷‍♂️

**Comment by hawkerm** [comment]
Quickly tried using `int` as the parameter type and then casting back to strongly typed:

```cs
    [RelayCommand]
    public async Task SwitchTheme(int newTheme)
    {
        ElementTheme = (MyLib.ElementTheme)newTheme;
        await ThemeSelectorService.SetThemeAsync(ElementTheme);
    }
```

This feels dirty, but works for now. Would be nice to keep strong typing through-out though.

**Comment by michael-hawker** [comment]
This seems related to this platform issue filed here: https://github.com/microsoft/microsoft-ui-xaml/issues/7633

**Reply by michael-hawker** [reply, marked as answer]
I have confirmed this is indeed a platform bug with UWP/XAML. I wrote a simple MarkupExtension and it provided the value and passed it correctly. I've posted that solution in that thread here: https://github.com/microsoft/microsoft-ui-xaml/issues/7633#issuecomment-1911242149

```cs
[MarkupExtensionReturnType(ReturnType = typeof(object))]
public class EnumValueExtension : MarkupExtension
{
    public Type Type { get; set; }

    public string Member { get; set; }

    protected override object ProvideValue()
    {
        return Enum.Parse(Type, Member);
    }
}
```

```xml
<RadioButton Command="{x:Bind ViewModel.SwitchPaneOrientationCommand, Mode=OneWay}"
             CommandParameter="{helper:EnumValue Type=model:PaneOrientation,
                                                 Member=HorizontalPreviewTop}"/>
```

**Reply by ZodmanPerth** [reply]
Works beautifully.  Thanks @michael-hawker.
